app/src/main/java/com/genymobile/gnirehtet/GnirehtetService.java

```diff
@@ -109,6 +110,19 @@ private void startVpn(VpnConfiguration config) {
         }
     }
 
+    @Override
+    public void onRevoke() {
+        try {
+            forwarder.stop();
+            forwarder = null;
+            vpnInterface.close();
+            vpnInterface = null;
+        } catch (IOException e) {
+            Log.w(TAG, "Cannot close VPN file descriptor", e);
+        }
+        stopSelf();
+    }
```

Comment:
Wouldn't it be possible to just call `close()`?

Reply:
stopSelf() - method from android api, iron destroy service.
I put close() method, service not stop immedly.